presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java

```diff
@@ -65,17 +82,34 @@ public LongDictionaryProvider(InputStreamSources dictionaryStreamSources)
      * @param items number of items expected in the dictionary.
      * @return The DictionaryResult contains two parts:
      * 1) the final dictionary buffer object. Different from the input dictionary buffer if the input
-     *    dictionary buffer is expanded or that the method returns a shared dictionary.
+     * dictionary buffer is expanded or that the method returns a shared dictionary.
      * 2) whether the caller will be the owner of the dictionary, for the purpose of memory accounting.
-     *    Callers own all non-shared dictionaries, and only the first caller of the shared dictionary
-     *    is the owner.
+     * Callers own all non-shared dictionaries, and only the first caller of the shared dictionary
+     * is the owner.
      * @throws IOException
      */
     public DictionaryResult getDictionary(StreamDescriptor streamDescriptor, long[] dictionary, int items)
             throws IOException
     {
-        InputStreamSource<LongInputStream> dictionaryDataStream = dictionaryStreamSources.getInputStreamSource(streamDescriptor, DICTIONARY_DATA, LongInputStream.class);
-        return loadDictionary(streamDescriptor, requireNonNull(dictionaryDataStream), dictionary, items);
+        InputStreamSource<LongInputStream> dictionaryDataStream = requireNonNull(dictionaryStreamSources.getInputStreamSource(streamDescriptor, DICTIONARY_DATA, LongInputStream.class));
+
+        if (streamDescriptor.getSequence() == 0 || dictionaryDataStream.openStream() != null) {
+            return requireNonNull(loadDictionary(streamDescriptor, dictionaryDataStream, dictionary, items));
+        }
+
+        // Try fetching shared dictionaries
+        DictionaryEntry entry = sharedDictionaries.get(streamDescriptor.getStreamId());
+        boolean isNewEntry = entry == null;
+        if (isNewEntry) {
+            StreamDescriptor sharedDictionaryStreamDescriptor = streamDescriptor.duplicateWithSequence(0);
+            InputStreamSource<LongInputStream> sharedDictionaryDataStream = dictionaryStreamSources.getInputStreamSource(sharedDictionaryStreamDescriptor, DICTIONARY_DATA, LongInputStream.class);
+            long[] dictionaryBuffer = loadDictionary(streamDescriptor, sharedDictionaryDataStream, dictionary, items).dictionaryBuffer();
```

Comment:
requireNonNull for sharedDictionaryDataStream.

Reply:
I could  add it, I'm slightly concerned because we already know this is never null, but it could be missingSource which returns null when opened, so the latter check is more important.

Reply:
I am thinking of the case, where there is a shared dictionary in flat map but the shared dictionary stream is missing in the file (due to writer bugs or file corruption)

Reply:
What I meant is that when the shared dictionary stream is indeed missing, the return value is a `missingStreamSource(streamType)`, which is only null when calling `openStream()`. We do have the check for when the stream is opened.

Reply:
I see makes sense, then, no check needs to be added.